### ISSUE TITLE
Multiline messages

### ISSUE THREAD

**Issue by highflyer77** [ircv3/ircv3-specifications#208]
CAP multi-line
This is only an incomplete example of how it could work, syntax or method to impliment would vary.

<lf>   would mark the new line (or whatever char. to be determined).
<crlf> would mark end of message, same as irc now uses.
- Raw message sent by user with multi-line support enabled:

```
<myNick>    This is my message I want to send. <lf>
            This would be line 2, I want to keep it intact with the first line. <lf>
            line 3 and end of the message. <crlf>
```
- How message "is" shown with "no" multi-line support:

`<myNick>   This is my message I want to send. This would be line 2, I want to keep it intact with the first line. line 3 and end of the message.`
- How message should be shown with multi-line support:

```
<myNick>    This is my message I want to send.
            This would be line 2, I want to keep it intact with the first line.
            line 3 and end of the message.
<OtherNick> Yeah cool.
```
- multi-line support would be usefull for pretty printing short code examples too.

```
<myNick>    #! /bin/bash
            echo "my cool script"
            if [ -n "$result" ] ; then
                echo "$result"
                exit 0
            else
                echo "File pattern not found"
                exit 1
            fi
<OtherNick> Nice code dude!
```
- The benefit is:  new lines are kept together without splitting up the message format.

I know some channels would have a limit or ban, but others could enjoy the multi-line support.
The 512 character limit would still apply to the message with multi-line support including the <lf> and <cr-lf> characters

Please comment and let me know if this is feasible,
Thank you.

**Comment by ChickenNuggers** [comment]
As I said on IRC, I think this would be better implemented via BATCH.

**Comment by dwfreed** [comment]
This wouldn't work in a lot of IRCds as they're currently implemented, because they accept LF from clients to mean the same thing as CRLF.

**Comment by ChickenNuggers** [comment]
> This wouldn't work in a lot of IRCds as they're currently implemented, because they accept LF from clients to mean the same thing as CRLF.

I know that, using [cqueues](https://github.com/wahern/cqueues/) for example, would not allow this; LF is automatically translated whether it be CRLF or LF coming in from the server.

**Comment by highflyer77** [comment]
> This wouldn't work in a lot of IRCds as they're currently implemented, because they accept LF from clients to mean the same thing as CRLF.

I sorta already knew that, that's why it says at the top:  " (or whatever char. to be determined). "

1. This was not meant to be used "as currently implemented" that's why it's being proposed.
2. If it's not implimented then they could just send multible messages, or ignore the LF, it's still not a deal breaker.
3. A CAP may not even be nessesary to impliment this.

It was only used for an example, to get the conversation started.  :-)

**Comment by kaniini** [comment]
Not ever happening in Charybdis.  Seriously, this is an over my dead body thing.

**Comment by MrBenC** [comment]
From IRC:
`<MrC> Other than spamming or ASCII art (both of which gets you banned and quickly in most places) I don't see it being that useful.  The code snippet example, really should be on a paste bin service.  I for one would not really enjoy having it scroll up as I'm trying to study it.`
`<MrC> And yes, consecutive lines by the same user could be displayed that way.  Think some clients actually do that.  But that should probably be up to the client.`

I'm sorry, but I really can't think of any actual legitimate use for it; the downsides aren't worth it.  To me this is just a means to prevent race conditions that may break up spam or ASCII art.

In the case of spam, could this prevent modes and kicks from say, a channel operator or bot that wants the spam to cease?  This doesn't seem like a good idea at all.

Ben

**Comment by jwheare** [comment]
The request here is entirely reasonable, the implementation can be figured out, probably using BATCH, it will need to be specified properly.

**Comment by dequis** [comment]
I think the main problem is that we'd like to have both multi-line messages and long payloads (#197), and both issues sound like they could be solved with BATCH but they overlap:

* If you have long messages as a batch that just concatenates them, you can't have multiple lines
* If you have a batch that makes each line a new line, you're still limited by (512-prefixes) for each message

**Comment by jwheare** [comment]
Long lines could perhaps be solved by a continuation tag. I don't see the conflict.

**Comment by jwheare** [comment]
I guess by "continuation tag" I just mean a different BATCH type. You can nest BATCH, e.g.

```
:irc.host BATCH +multi123 multiline
@batch=multi123 :nick!user@host PRIVMSG #channel :Line 1
@batch=multi123 :irc.host BATCH +cont123 continuation
@batch=cont123 :nick!user@host PRIVMSG #channel :Line 2...512 limit reached
@batch=cont123 :nick!user@host PRIVMSG #channel :Line 2 continued
@batch=multi123 :irc.host BATCH -cont123
@batch=multi123 :nick!user@host PRIVMSG #channel :Line 3
:irc.host BATCH -multi123
```

**Comment by ChickenNuggers** [comment]
@dequis The point of this is that you don't have to worry about the 512 limit except for how normal lines work, which makes it entirely compatible with non-IRCv3. @jwheare's example explains this better, I guess.

**Comment by SaberUK** [comment]
:-1: for introducing special message parsing rules but :+1: for making it a batch type.

**Comment by ChickenNuggers** [comment]
I believe we can uniformly agree against special message parsing rules.

**Comment by highflyer77** [comment]
I can't see any harm with parsing rules.  

 1. If the existing clients that don't support it, would just simply just ignore it, no problems for anyone.  
Or
 2. Would just simply just create multiple messages, still no problems.  

Only the clients that support it would benefit. 
But # 1. and # 2. would depend on how the parsing rules are implimented.  

* This is easier to implement, it has lower overhead required compared to sending batch especially for simple short text.

It would also co-exist with batch, so having the best of both could be possible.

**Comment by ChickenNuggers** [comment]
* BATCH exists.
 * BATCH works.
 * BATCH would require no further work implementing parsing rules.

**Comment by highflyer77** [comment]
@ChickenNuggers   
The spec says:
> Client may show the messages on the channel in order 1,2,3,4,5, or in order 1,4,2,3,5, among others.

**That doesn't help keep the message formated.**
Also, what happens when there is no ending batch tag?
Like if someone didn't finish or got disconnected, etc.  
I can foresee a lot of problems.

**Comment by SaberUK** [comment]
>If the existing clients that don't support it, would just simply just ignore it, no problems for anyone.

I think you are significantly overestimating the standard compliance of client message parsers.

**Comment by highflyer77** [comment]
@SaberUK 
I meant that the client parser would just ignore the char. as of "now".
(Without the client changing a thing, as of today or in the future).  
The char. would just not get used for the newline, so no problems for anyone, but the new clients could support it, if they wanted to.

**Comment by SaberUK** [comment]
>I meant that the client parser would just ignore the char. as of "now".

I repeat what I said, I think you are significantly overestimating the standard compliance of client message parsers.

**Comment by dequis** [comment]
A decent client parser will just handle cr, lf and crlf as valid line endings, because you can't trust the server to send crlf. Same thing from the point of view of servers. Ensuring interoperability with as many implementations as possible is often a higher priority than implementing exactly what the rfc says.

Allowing clients to send linefeeds embedded in messages is equivalent to allowing them to send raw commands.

**Comment by ChickenNuggers** [comment]
> Also, what happens when there is no ending batch tag?

@highflyer77:

Then the server sends one to the clients, realizing the client disconnected, or discards the message entirely.

**Comment by ChickenNuggers** [comment]
Would anyone support this in their software? If so, I'll draft up a spec for it using BATCH.

**Comment by ChickenNuggers** [comment]
(which assumes that BATCH Client to Server is implemented when it's accepted)

**Comment by highflyer77** [comment]
* Using this method of "multi-line" support doesn't require the server to do any "work" to impliment it.  

(Like sending a ending batch tag or waiting for a timeout, etc. and we still have the ordering problem that was listed above w/batch).  

Using delimiters, the only thing the server would need to do is:  
 1.  Show "multi-line" support. 
 2.  Show which **"char."** it wants to use for the for the new line delimiter (for the parsing of messages), 
	but maybe this could be sent with the message that is using the "multi-line" support (like a color code does, but except it would be for a newline).
 
**Everything else would be done on the client side.**  The client would do all the parsing if they choose to support it.
	If the client doesn't support "multi-line", it would just simply show the entire message, including the delimiter chars., that were sent with the message.

@ChickenNuggers 
You keep saying you are going to draft up a spec. for your BATCH method, but I haven't seen anything yet.
And, why don't you post your ideas on your own BATCH issue, instead of constantly spamming/bashing mine, and everyone elses opinions?
Can't you just make a more unified formal opinion, instead of so many dang (spams) ?

**Comment by kaniini** [comment]
Again I feel I should remind people for the sake of conversation that it is technically infeasible to implement this in IRCd, as IRCd needs to be compatible with `\r`, `\n` and `\r\n` line-endings.  This behavior is required by RFC1459 (for full conformance).

While not all IRCds treat those line-endings equally, this alters the message parsing semantics in a way that is certainly in no way backwards compatible with RFC1459, and would in many cases (DBuf/linebuf based IRCd which is basically all of them except InspIRCd and ngIRCd) require significant modification to the way the DBuf/linebuf code handles framing.

**Comment by highflyer77** [comment]
@kaniini 
If you re-read the last message.  I say again:
>Everything would be done on the client side. The client would do all the parsing if they choose to support it.

**No line-endings need to be changed with the server**

**Comment by SoniEx2** [comment]
[This is already a thing with a "compliant" (whatever that means) CTCP implementation.](http://www.irchelp.org/irchelp/rfc/ctcpspec.html)

**Comment by kaniini** [comment]
@highflyer77 why does it need to be a cap at all then?

**Comment by highflyer77** [comment]
@kaniini 
As I also said in my second message above:
> 3 .  A CAP may not even be nessesary to impliment this.

**Comment by kaniini** [comment]
how will you handle clients which do not support your changed semantics?  will they just have random control codes in the text?  sounds like a shitty user experience.

**Comment by highflyer77** [comment]
@kaniini 
Thank you, for your kind comments.  
It would only be a "shitty user experience" if **you** choose to make it so.  
It could be any normal char. like:   ";" or  "|"  or
It could be any uni-code char. like a: ⏎ or any shitty char. your heart desires.
But seriously, they don't have to be "random control codes" or "invisable chars.", they could be anything you wish.  
So the clients that **don't** support "multi-line" support, might look like this:  
`<myNick>     This is my message I want to send. ⏎ This would be line 2, I want to keep it intact with the first line. ⏎ This would be line 3 and end of the message.`

In fact showing the return char. of " ⏎ " might be a good thing, because then the user would know if multi-line support isn't enabled. But like I said, it could be any character used as a delimiter.

All suggestions and comments are welcome, except negative ones. Haha, just joking kaniini.
Thanks

**Comment by jwheare** [comment]
So then that character becomes unusable in normal text without escaping. It's not workable. If you want hacks like this, use CTCP.

**Comment by ChickenNuggers** [comment]
> It could be any normal char. like: ";" or "|"

You say this, but you also said that you should use this for code. I've seen multiple languages that use both.

**Comment by highflyer77** [comment]
@ChickenNuggers 
I'm saying you can use any character your heart desires, why nitpick?

**Comment by kaniini** [comment]
Can we just close this already?  When somebody has something concrete based on BATCH, we can revisit.

**Comment by ChickenNuggers** [comment]
> 11:47 <dx> yes, i'd like multi-line messages too, but batches seem to be preferred as the way to do it
11:47 <strayPuppy> well then there you go, if everyone agreed , that's off the table.

Confirmed that this should use BATCH.

**Comment by highflyer77** [comment]
And the next line says:
> but I don't know if everyone agrees, and it hasn't been fully discussed yet either. 

I hate nitpickers and spammers and seems a lot of it going around by lately.

1. I think there are PRO's and CON's to each method and there hasn't been any up or down vote on anything, so I'd say that Nothing is: "Confirmed". 
2. I haven't even seen any up or down vote on the idea of "multi-line" yet.

When I see the specs. I'll draw up a list of Pro's and Con's for both, then the decision can be made on one, the other, both or none.

**Comment by MicroDroid** [comment]
Well, I think this is going to be like, **Make `^L` (0xC) act as newline character**
This would also need to be `CAP`ed
The procedure is like:
# For a server

- Strip out all `^L`s when sending messages to some client that doesn't support multi-line
- Otherwise send like any message
- Do not allow more than **n** (Like, 5 for ex) newlines to flow per message

# For the client

- If server supports `multi-line`, then replace all user-inserted newlines with `^L` when sending them
- As well as convert back `^L`s to newlines once they are received
- If the server does not support `multi-line`, strip out newlines
- The client is **RECOMMENDED** to keep replacing `^L`s with newlines even if the server does not support it, as the user my forcibly send a `^L`
- If the total message length is more than **512 bytes**, then the client is supposed to replace the `^L` with a real `CRLF` and send two messages, as it does now.

I can't see why this would be conflicting with backward compatibility or anything else.
Also messages still should be limited to **512 bytes** to prevent instant spamming, like sending a kilometer long message at once

**Comment by jwheare** [comment]
Special control characters have been ruled out for this. Again, this will need to be specified as a BATCH to gain any further approval.

**Comment by MicroDroid** [comment]
Well assuming by now we have a stable client-sent BATCH thing, how about:

# For the server

Here's how you handle old clients:

- Combine the BATCHs into several **PRIVMSG**s when sending them to the client **(I recommend this approach as the client still receives a multi-line message)**

And here's the flow with the supported clients:

- Upon connection, send a BATCH limit, like, the limit of allowed open BATCHs at once
- When receiving a BATCH, if its to a user, pass with no line-limit
- If the BATCH is for a channel, read BATCH commands one by one, if they don't reach the **PRIVMSG/NOTICEs per BATCH** limit, process as normal and broadcast to clients
- Otherwise, forcibly end the BATCH, and send a `MESSAGE_TOO_LONG` (or whatever) refusal response with the BATCH reference tag

# For the client

To handle receiving messages:

- If the server does not support the capability, just replaces newlines with nothing, or like, do whatever the client did before that capability existed
- Receiving a multi-line message is like receiving a  BATCH with only PRIVMSGs, if the BATCH contains anything other than BATCHs, then treat as separate normal IRC command.
- Now, if the BATCH is a multi-line message simply combine all the messages into one and display to the user

And here's how would a client handle sent messages:

- If a **PRIVMSG** contains no newlines, send as a normal PRIVMSG
- If the open BATCHs are to the server limit, then stack the message on a queue
- A supported server is to send a `CHANNEL_BATCH_LIMIT`, which when sending BATCHs, check against that limit, if the sent-message contains too many lines, send them on a next BATCH
- If it does contain newlines, then start a new BATCH and..
- Take the message as a `string`, split that string to an array, with `\n` being the delimiter, loop over each part, if the *part + the command* exceed **512 bytes** then split that message somewhere, while possibly preserving the full word (i.e. don't split 'Welcome to my home' to ['Welcome to my ho', 'me'], but rather ['Welcome to my', 'home']
- Now send all that parts on the started BATCH
- End the batch

With this, we would be able to have a balance between preventing spamming, a channel op control, and yet allowing multi-line messages to be sent
By the way the server may *refuse* the messages and not even send a part of them, as the right client should maintain that limit anyways

**Comment by attilamolnar** [comment]
I'd like this to be implemented by making use of a better wire format (such as the proposed json or ircstrings format). The hackery with BATCH is ugly because it is working around a low level problem (wire format not allowing specific chars) by abusing a higher level construct designed for indicating related but distinct events (not for splitting one single event into multiple ones).

**Comment by MicroDroid** [comment]
*shrugs*

**Comment by ozjd** [comment]
Several older clients (eg. MS Comic Chat) had implement multi-lines by using the following characters as a separator:
Data Line Escape (0x16),  'r' (0x72), Data Line Escape (0x16), 'n' (0x6E)

Supporting this kind of thing should be client implementation IMHO.

**Comment by attilamolnar** [comment]
As of now one possibility to implement it is by using client to client tags. For example: send messages as usual and whenever the next message is the continuation of the previous one, tag it. Clients that understand the tag can show the messages as one.

**Comment by ChickenNuggers** [comment]
Example:

```
ryan!admins@hashbang.sh PRIVMSG #testing :this is a
@+multiline ryan!admins@hashbang.sh PRIVMSG #testing :test of multiple
@+multiline ryan!admins@hashbang.sh PRIVMSG #testing :lines
ryan!admins@hashbang.sh PRIVMSG #testing :this is a single line
```

@attilamolnar you mean something like this, correct?

**Comment by jwheare** [comment]
Probably would make sense to add a unique id too, like `@+multiline=abdef123` so you could send several separate multilines in a row.

Receiving clients would probably need to implement some sort of time or number of message based threshold for appending lines to the group, in case a sending client decided to reuse ids later or if they get interleaved with other messages.

**Comment by ChickenNuggers** [comment]
I was hoping it could be more linear honestly, but if we have an asynchronous receiving system, that's fine.

**Comment by dequis** [comment]
I think it would be better to send the tags in every message except the last, instead of every message except the first. Derived from the example above:

```
@+multiline ryan!admins@hashbang.sh PRIVMSG #testing :this is a
@+multiline ryan!admins@hashbang.sh PRIVMSG #testing :test of multiple
ryan!admins@hashbang.sh PRIVMSG #testing :lines
ryan!admins@hashbang.sh PRIVMSG #testing :this is a single line
```

This helps clients telling them to hold a reference to that point, if the text buffer lacks metadata to find messages by nickname after the fact. Would also help scripts/bots that wish to handle multiple lines, telling them to not process the message just yet.

>Probably would make sense to add a unique id too, like `@+multiline=abdef123` so you could send several separate multilines in a row.

I think it's enough to not add any tags in the last message

**Comment by ChickenNuggers** [comment]
> I think it's enough to not add any tags in the last message

The issue isn't with saying the end as much as it is with supporting multiple multi-line messages (which I'd rather not personally)

I do like your idea for removing the tag at the end. :+1:

**Comment by louiz** [comment]
So, I didn’t really understand why this issue stalled, and there’s no proposal like this

```
BATCH +abcd multiline
@batch=abcd int main()
@batch=abcd {
@batch=abcd    return 0;
@batch=abcd }
BATCH -abcd
```

(the important part would be the “multiline” type, which specifies that the client SHOULD display that batch as one single multi-lines message)

Is there still no consensus on this way of doing things, or are we just waiting for a good will to write this proposal?

**Comment by Zarthus** [comment]
Issues tend to held back by nobody contributing (and maintaining) a pull request in the form of the specification.

**Comment by dequis** [comment]
@louiz Batches have indeed been mentioned multiple times in this thread, but the main issue with those is that there's no client-side batch verb in the batch-3.2 spec, and adding a verb like that has plenty of awkward implications - should the server buffer batches? should the server track them? And so on.

The answer to those questions usually ends up with the conclusion that the involvement of servers in "client batches" needs to be minimal, just blindly relaying them. Which ties into the existence of client-to-client message tags, which were not a thing for most part of this discussion.

Basically, consider all the discussion before attila's comment in https://github.com/ircv3/ircv3-specifications/issues/208#issuecomment-259956692 to be obsolete.

**Comment by MuffinMedic** [comment]
As was an issue with https://github.com/ircv3/ircv3-specifications/pull/285#discussion_r94814237, would using client generated `@+multiline=ID` introduce issues of guaranteed uniqueness?

**Comment by jwheare** [comment]
Expanding on an IRC discussion the other day, I can see a few options.

Here's an example message we're trying to send:

```
start of a concatted message [...snip...] exceeds 512 chars rest of the line
continuing the message after a new line
another new line and the end of the message
```

In these examples `concat` means the line was split for being too long, and `concat-line` indicates a line break.

In order of most to least robust (imo):

---

1) Use msgids and allow a client to send an entire batch to the server. The server wouldn't necessarily have to buffer the input, just keep track that it's in a batch and not assign msgids to inner messages.

```
Client sends:
BATCH +asdf concat
@batch=asdf :prefix PRIVMSG #chan :start of a concatted message [...snip...] exceeds 512 chars
@batch=asdf :prefix PRIVMSG #chan :rest of the line
@batch=asdf;concat-line :prefix PRIVMSG #chan :continuing the message after a new line
@batch=asdf;concat-line :prefix PRIVMSG #chan :another new line and the end of the message
BATCH -asdf

Server adds a msgid to the start of the batch and relays:
@msgid=123 :irc.example.com BATCH +asdf concat
@batch=asdf :prefix PRIVMSG #chan :start of a concatted message [...snip...] exceeds 512 chars
@batch=asdf :prefix PRIVMSG #chan :rest of the line
@batch=asdf;concat-line :prefix PRIVMSG #chan :continuing the message after a new line
@batch=asdf;concat-line :prefix PRIVMSG #chan :another new line and the end of the message
:irc.example.com BATCH -asdf
```

---

2) Use a mechanism that lets the server inject msgids given a start and end tag. This requires some state to be tracked server side, but no message buffering. It's up to the client to piece the parts together.

The tracking would be: receive `concat-start`, assign a msgid for the client and store it in internal state, add it to subsequent `concat` / `concat-line` messages until `concat-end` then discard. Optionally discard if messages arrive without any `concat` tags or on a timeout.

```
Client sends:
@concat-start :prefix PRIVMSG #chan :start of a concatted message [...snip...] exceeds 512 chars
@concat :prefix PRIVMSG #chan :rest of the line
@concat-line :prefix PRIVMSG #chan :continuing the message after a new line
@concat-line;concat-end :prefix PRIVMSG #chan :another new line and the end of the message

Server relays:
@msgid=123;concat-start :prefix PRIVMSG #chan :start of a concatted message [...snip...] exceeds 512 chars
@concat=123 :prefix PRIVMSG #chan :rest of the line
@concat-line=123 :prefix PRIVMSG #chan :continuing the message after a new line
@concat-line=123;concat-end :prefix PRIVMSG #chan :another new line and the end of the message
```

---

3) Client uses client tags, sends the first line and waits for an echo message to discover the msgid and adds it to the other lines itself. This relies on the client validating that the same user is concatting their own messages, by comparing the nick and making sure it hasn't changed hands via NICK or JOIN between start and end. The server will add unique msgids on each message but they don't correlate.

```
Client sends:
@+concat-start :prefix PRIVMSG #chan :start of a concatted message [...snip...] exceeds 512 chars

Server relays and echoes:
@msgid=123;+concat-start :prefix PRIVMSG #chan :start of a concatted message [...snip...] exceeds 512 chars

Clients sends rest:
@+concat=123 :prefix PRIVMSG #chan :rest of the line
@+concat-line=123 :prefix PRIVMSG #chan :continuing the message after a new line
@+concat-line=123;concat-end :prefix PRIVMSG #chan :another new line and the end of the message

Server relays rest as is:
@msgid=456;+concat=123 :prefix PRIVMSG #chan :rest of the line
@msgid=789;+concat-line=123 :prefix PRIVMSG #chan :continuing the message after a new line
@msgid=abc;+concat-line=123;+concat-end :prefix PRIVMSG #chan :another new line and the end of the message
```

---

4) As in 3, but do ignore the message IDs entirely, don't wait for an echo, and have receiving clients walk back through messages comparing the nickname as before. This is easier for the client to send, but a bit less robust to piece together without ids.

```
Client sends:
@+concat-start :prefix PRIVMSG #chan :start of a concatted message [...snip...] exceeds 512 chars
@+concat:prefix PRIVMSG #chan :rest of the line
@+concat-line :prefix PRIVMSG #chan :continuing the message after a new line
@+concat-line;concat-end :prefix PRIVMSG #chan :another new line and the end of the message

Server relays:
@msgid=123;+concat-start :prefix PRIVMSG #chan :start of a concatted message [...snip...] exceeds 512 chars
@msgid=456;+concat :prefix PRIVMSG #chan :rest of the line
@msgid=789;+concat-line :prefix PRIVMSG #chan :continuing the message after a new line
@msgid=abc;+concat-line;+concat-end :prefix PRIVMSG #chan :another new line and the end of the message
```

---

Other whackier ideas might involve enabling a server to accept much longer lines and split them itself, and treating a `\n` as non message ending (as opposed to `\n\r`.

**Comment by jwheare** [comment]
I've implemented mechanism 2 in the IRCCloud Slack integration, but I'll give mechanism 1 with BATCH a go to see if it sits better.

**Comment by jesopo** [comment]
Just going to throw my two cents in here by echoing something @RyanSquared said.

https://github.com/ircv3/ircv3-specifications/issues/208#issuecomment-180060056
> * BATCH exists.
> * BATCH works.
> * BATCH would require no further work implementing parsing rules.

There's currently no precedent for client->server `BATCH`es but I really think the backwards-compatibility of putting a bunch of `PRIVMSG`s in a `BATCH` makes this a winner for me. We've already speced batches out and they're designed for graceful fallback.

**Comment by SaberUK** [comment]
Using C2S batches for multi-line messages means the server needs to have special handling for batched messages when realistically the server will never care about this.

Is there any reason client-to-client tags can't be used for this purpose instead?

**Comment by SaberUK** [comment]
I'm opposed to C2S batches until I can find an actual use case for them where the server might actually care about batching messages. See my comment in #208 for why I don't think they are suitable for multi-line messages.

**Comment by nomis** [comment]
There's a lot of resource overhead for servers to control `BATCH` from clients (across multiple channels) and protocol complications when you have multiple servers (do you relay before the batch finishes? how does that affect filtering? do all servers on the network then have to queue all batches?).

The same issue exists on the client side.

Any kind of line combining presents a filtering problem on servers that want to do that (e.g. sending `h`, `t`, `t`, `p` separately). We need to be clear whether this issue is for multiple lines or also for concatenating multiple messages into one or more lines.


I think it would be a lot simpler to use a client tag to associate each message that is part of a multi-line message with a unique identifier. Multiple lines can be merged into one block (without concatenating lines) by the client by comparing each message with the previous message sent by that user in the channel.

Clients then only need to store one message identifier per channel user and an internal reference to where that message is in the display buffer. The start timestamp of the sequence of messages and the timestamp since the previous message can be used to determine whether or not it is acceptable to merge all the lines into one display block over a period of time.

**Comment by jesopo** [comment]
If this was to be implemented as c2c, surely it would be able to bypass spam filtering? unless you want the server to get in on the action - in which case why not just use a batch so the server knows where the data ends and can correctly spamfilter it?

**Comment by SaberUK** [comment]
>If this was to be implemented as c2c, surely it would be able to bypass spam filtering?

Servers can look at client-to-client tags too if they really want to.

>unless you want the server to get in on the action - in which case why not just use a batch so the server knows where the data ends and can correctly spamfilter it?

The server can do this with client-to-client tags without the overhead of c2s batches.

**Comment by jesopo** [comment]
how does the server retract a half-formed malicious line when it's already sent?

**Comment by grawity** [comment]
> how does the server retract a half-formed malicious line when it's already sent?

I suggest `:foo.example.com @cancel=<messageid> TAGMSG #chan`

I think that's similar to how Twitch's IRC gateway handles message deletions.

It won't work with old clients lacking multiline support, but at least the server can prevent the message from being _fully_ sent, so all that everyone else sees is a truncated unclickable URL or similar. IMHO isn't much worse than the current situation.

(I'd probably implement some fancy spamfilter rule that bans or disconnects the user if the message matches a spamfilter but is far too short to have actually _needed_ multiline wrapping... e.g. someone sending it in 10-char chunks clearly does so just to bypass the filtering.)

**Comment by jesopo** [comment]
So... you're saying you can't retract the message from non-compat clients?

**Comment by grawity** [comment]
No, I'm saying you can't retract the part of message that has been sent prior to spamfilter detection from non-compat clients.

**Comment by jesopo** [comment]
so non-compat clients are going to see half-message artifacts but compat clients aren't? please keep backwards compatibility in mind here.

**Comment by grawity** [comment]
> so non-compat clients are going to see half-message artifacts but compat clients aren't?

Yes. (Which, frankly, is at worst no different from the current situation where someone pastes an ASCII art bomb and the spamfilter kicks in maybe 4–5 lines into the message, and you cannot retract the already-sent lines either.)

> please keep backwards compatibility in mind here.

This shouldn't even affect the majority of legitimately line-wrapped messages, because they either don't get spamfiltered in the first place (therefore no retraction), or are sent in full-size chunks and the spamfilter matches a whole line at once. The latter, again, is no different from the current situation of kicking someone in the middle of a multi-message burst.

I don't think backwards compatibility is a significant issue when we're discussing spam messages which are _deliberately_ meant to disrupt the chat... , does it? (And if it does – as above – it's still not really worse than the current situation.)

**Comment by DanielOaks** [comment]
I mean, the breaking compat thing is mostly around presenting protocol stuff that older clients can't handle it, as well. I don't consider @grawity's suggestion to be tooo bad there. If a server really wanted to, they could also cache multi-line messages until they receive the full thing and then dump it in one go after completing the spam-filtering, which would work fine.

**Comment by jesopo** [comment]
> Yes. (Which, frankly, is at worst no different from the current situation where someone pastes an ASCII art bomb and the spamfilter kicks in maybe 4–5 lines into the message, and you cannot retract the already-sent lines either.)

I remind you we're trying to make things better.

> I don't think backwards compatibility is a significant issue when we're discussing spam messages which are _deliberately_ meant to disrupt the chat... , does it? (And if it does – as above – it's still not really worse than the current situation.)

It does when it leads to inconsistent data across clients 

> If a server really wanted to, they could also cache multi-line messages until they receive the full thing and then dump it in one go after completing the spam-filtering, which would work fine.

Perfect case for batches surely?

**Comment by jesopo** [comment]
What is the "overhead" of batches, by the way? how taxing can they really be?

**Comment by SaberUK** [comment]
>how does the server retract a half-formed malicious line when it's already sent?
>> unless you want the server to get in on the action - in which case why not just use a batch so the server knows where the data ends and can correctly spamfilter it?

>The server can do this with client-to-client tags without the overhead of c2s batches.

**Comment by jesopo** [comment]
Batches don't **have** to be either.

**Comment by SaberUK** [comment]
>Batches don't **have** to be buffered either, they're just better suited to being able to be.

What exactly is the benefit of usi

I'm yet to see an argument that actually involves using client to server batches that doe

**Comment by SaberUK** [comment]
>Batches don't **have** to be buffered either, they're just better suited to being able to be.

What exactly is the benefit of adding client-to-server batches over using continuation tags? The primitives already exists for doing this with client tags and they require very little overhead.

**Comment by jesopo** [comment]
Where are those primitives? What's the server overhead of a batch?

Really I just don't want to have to completely duplicate my batching logic to batch by a message-tag. `BATCH` makes this spec-defined as a generalised method of message relating - why should we not use this, why should we duplicate a spec and why should we duplicate code?

**Comment by jwheare** [comment]
Following some chit chat in IRC, some of us are gonna take a look at the implementation issues in code before getting more stuck in the weeds and/or entrenched.

**Comment by DarthGandalf** [comment]
> If a server really wanted to, they could also cache multi-line messages until they receive the full thing and then dump it in one go after completing the spam-filtering, which would work fine.

Surely spammers won't send their spam as a single multiline message, but in several messages instead. Just like they do now.

**Comment by jesopo** [comment]
I've somewhat changed my opinion on this over time

Although I think a client->server batch would be the **neatest** way to do this, I now lean towards just using message-tags on their own for it because having to introduce client->server batches **just for this** is much too much to work out and much too much to come to an agreement on.

**Comment by eeeeeta** [comment]
As a halfway-house decision, maybe using c2s message-tags for sending (similar to @jwheare's method 2 from 
https://github.com/ircv3/ircv3-specifications/issues/208#issuecomment-285516349) and then s2c batches for relaying may be the best of both worlds?

c2s interface:
- pretty damn easy for clients to implement (just tack some `@concat-*` tags onto the messages)
- _does_ require the server to special-case concat message sending and track these tags, which might be a point of contention

s2c interface:
- uses batches, so presumably easy to add support for if you already do batches
- allows the server to solve the problem of aborting messages mid-send, as they keep a cache for every user and only release the batch on concat-end
- doesn't match up with how the messages were actually sent, but maybe this isn't a problem

**Comment by jesopo** [comment]
Got an implementation of the client->server `BATCH` option for this in BitBot as `bitbot.dev/multi-line`: https://github.com/jesopo/bitbot/blob/5fa3d66a39dc22bff9a1d0345ebb70448b060e82/modules/ircv3_multi_line.py


Was simple and neat to implement, quite happy with it.

**Comment by jesopo** [comment]
Some notes on issues with how this interacts with other specs.

- `echo-message` will need to echo the whole `BATCH` back
- `labeled-response` will either have to wrap the `BATCH` in a `labled-response` BATCH or put the `label` tag on `BATCH +123 multiline`
- `msgids` can easily just put the msg id on the `BATCH` as it is echoed and relayed but this degrades when being sent to legacy clients that support `msgid` but not `multiline`. Does the `msgid` tag go on only the first line in the `BATCH`? does each message get a different `msgid` (maybe `123-1` `123-2` `123-3`?)
-  `editmsg` should work fine if the `msgid` issue is resolved.

**Comment by jesopo** [comment]
solutions:
- `labeled-response`: `@label=123 BATCH +123 multiline`
- `msgids`: msgid on relayed `BATCH +` for clients that support `batch`, msgid on first line for clients that do not

**Comment by jwheare** [comment]
Closing this issue in favour of the batch proposal #398 for now.

**Comment by minecraftchest1** [comment]
Hope I am not to late to the discussion, but here is my 2 cents anyways.
A multi line tag could have an ID generated based on the time that the first message was created. This would help keep the ID's unique, and would reduce the required workload on the server.
If there are any clients that don't support the tag connected, the server should then create a batch for the message and send it to that client, and send the tag to the rest of the clients.

**Comment by jwheare** [comment]
That’s what message ids are for 

https://ircv3.net/specs/extensions/multiline#message-ids-spec
https://ircv3.net/specs/extensions/message-ids

**Comment by minecraftchest1** [comment]
I should've mentioned that I found this issue by searching for IRC newlines in message, and haven't read the specs yet.

Sent from my T-Mobile 4G LTE Device
Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Comment by a19102618** [comment]
I'm trying to send multiline messages to an irc server, but all messages are sent individually by the server. Is the multiline support implemented in any irc server?

**Comment by classabbyamp** [comment]
see https://ircv3.net/support/networks. did you request the cap and did the server acknowledge it?

**Comment by a19102618** [comment]
@classabbyamp thanks, it works on networks that have implemented CAP draft/multiline.
